tasks/Task.cpp

```diff
@@ -28,27 +27,289 @@ bool Task::configureHook()
+        task.set_cVe(cVe);
+        task.print();
+        std::cout << std::endl;
+        std::cout << "cdMo" << std::endl;
+        poseDisplay(cdMo);
+        std::cout << "cMo" << std::endl;
+        poseDisplay(cMo);
+
+        v = task.computeControlLaw() ;
+
+        writeVelocities(v);
+        corners.clear();
+    }
 }
+
 void Task::errorHook()
 {
     TaskBase::errorHook();
 }
+
 void Task::stopHook()
 {
     TaskBase::stopHook();
 }
+
 void Task::cleanupHook()
 {
     TaskBase::cleanupHook();
+    
+    task.kill();
 }
+
+void Task::updateFeatures(std::vector<base::Vector2d> corners)
+{
+    //  Compute the initial pose
+    pose.clearPoint();
+
+    for (int i=0 ; i < 4 ; i++)
+    {
+        base::Vector2d aux;
+        aux = corners[i];
+        //vpImagePoint is instanciated using (v,u) instead of (u,v)
+        vpImagePoint a(aux[1],aux[0]);
+        double x,y;
+        vpPixelMeterConversion::convertPoint(cam_cal,a,x,y);
+        point[i].set_x(x);
+        point[i].set_y(y);
+        pose.addPoint(point[i]) ;
+    }
+
+    pose.computePose(vpPose::LAGRANGE, cMo);
+    pose.computePose(vpPose::VIRTUAL_VS, cMo);
+
+    // Update Feature 1 - p
+    // Sets the current position of the visual feature
+    P.track(cMo);
+    vpFeatureBuilder::create(p, P);
+
+    //////////////////////////////
+    // Update Feature 2 - depth //
+    //////////////////////////////
+
+    //Takes the features, calculate the intersection of the diagonals of the square and
+    //given the camera parameters, retrieve the coordinates in meters.
+    vpImagePoint refP;
+    //double x,y;
+    vpImagePoint cog[4];
+
+    for (int i=0; i< 4; i++)
+    {
+        base::Vector2d aux;
+        aux = corners[i];
+        //according to vpImagePoint documentation (i,j) is equivalent to (v,u)
+        cog[i].set_ij(aux[1], aux[0]);
+    }
+
+    // Compute the coordinates of the 2D reference point from the
+    // intersection of the diagonals of the square
+    double a1 = (cog[0].get_i()- cog[2].get_i()) /
+        (cog[0].get_j()- cog[2].get_j());
+    double a2 = (cog[1].get_i()- cog[3].get_i()) /
+        (cog[1].get_j()- cog[3].get_j());
+    double b1 = cog[0].get_i() - a1*cog[0].get_j();
+    double b2 = cog[1].get_i() - a2*cog[1].get_j();
+
+    refP.set_j( (b1-b2) / (a2-a1) );
+    refP.set_i( a1*refP.get_j() + b1 );
+    double x,y;
+    vpPixelMeterConversion::convertPoint(cam_cal, refP, x, y);
+
+    // Update the 2D ref point visual feature
+    double Z;
+    Z = P.get_Z();
+    p.set_xyZ(x, y, Z);
+    depth.buildFrom(P.get_x(), P.get_y(), Z, log(Z/Zd));
+
+    //////////////////////////////
+    // Update Feature 3 - t, tu //
+    //////////////////////////////
+    cdMc = cdMo*cMo.inverse();
+    t.buildFrom(cdMc);
+    tu.buildFrom(cdMc) ;
+
+}
+
+void Task::updateDesiredPose()
+{
+    // Initialise a desired pose to compute s*, the desired 2D point features
+    base::LinearAngular6DCommand setpoint;
+    _cmd_in.read(setpoint);
```

Comment:
You have *zero* guarantee that there is a sample on this port, especially when you call updateDesiredPose in the startHook